### DISCUSSION TITLE
Change the mac config path from ~/Application Support/nushell/config.nu to ~/.config/config.nu

### DISCUSSION THREAD

**Discussion by norman784** [nushell/nushell#5279] [Ideas]
I don't know why it was build this way, but I find really odd to have my config files outside the ~/ or ~/.config folder, I usually have one set of dot files that use between my mac and linux machines (currently not using linux, but from time to time I use it and I like to have a portable set of dot files), so for me having the config.nu in another directory throws me off.

**Edit:** It seems that the consensus is to use the `~/Application Support` folder because the XDG spec is linux only (while every cli app I used, except nushell, uses it on macOS).

**Comment by kubouch** [comment]
I think we use a crate to give you the default config directory based on your OS. Perhaps we could tweak it, I'm not sure what the default is supposed to be on MacOS.

However, you still have a couple of options:
* symlink `~/.config/config.nu` to `~/Application Support/nushell/config.nu`
* source `~/Application Support/nushell/config.nu` within `~/.config/config.nu`
* launch Nushell with `nu --config "~/Application Support/nushell/config.nu"`

**Reply by norman784** [reply]
For desktop apps the `~/Application Support` folder is used, but most terminal apps just use `~/.config` folder as linux.

I'm already doing the source to the `~/.config` folder, but when I setup a new user for example I need to remember to follow those steps, and because that's not something you do everyday, you might forgot to do so.

**Reply by flying-sheep** [reply]
The XDG spec is Linux only. `Application Support` is correct on macOS.

**Reply by yamiteru** [reply]
I've created a simple shell script that copies config from `~/.config/nushell` to Applicatio Support and I manually run it after every change.

**Comment by fdncred** [comment]
I'm pretty sure this is the crate we're using https://github.com/xdg-rs/dirs/blob/af4aa39daba0ac68e222962a5aca17360158b7cc/directories/src/mac.rs#L10

**Comment by eugenesvk** [comment]
I'd also prefer the shell not to use ~/Application Support as that's very anti-shell and (besides the reasons of portability) doesn't play nice with other tools that, for example, backup user config, but only from $HOME and would require inconvenient simlinky workarounds to work with ~/Application Support which are sometimes broken by the tools themselves when they decide to do some symlink-unaware update of their working folder

There is a good suggestion that might satisfy both the followers of the official path and those preferring the cross-platform convenience:
>https://github.com/xdg-rs/dirs/issues/45#issuecomment-1090308036
> respect the XDG_CONFIG_HOME env var if set and fall back to the default macOS dir if not.

**Reply by flying-sheep** [reply]
macOS has its standard directories, using some standard that exists for another platform makes no sense.

Nu could simply grow a `nu --paths` option or `nu path` command to show where things are.

**Reply by fdncred** [reply]
`$nu` shows where config files are now.

**Reply by flying-sheep** [reply]
Perfect!

**Reply by eugenesvk** [reply]
> macOS has its standard directories, using some standard that exists for another platform makes no sense.

What makes no sense is you sticking to some OS standard just because it exists without any justification. In my comment and the original topic it is explained why deviating from that standard makes sense — it improves our cross-platform lives. And this non-standard location isn't forced on anyone — a user has to opt in, so what's the downside? If you like the standards — you'll get them by default!

**Reply by flying-sheep** [reply]
You’re probably right. People define the XDG variables are manually so someone wanting to use the macOS standard dirs will either not have them defined or will have them point to the macOS standard dirs.

It’s hard to anticipate this behavior though. If everything would use standard dirs, everyone’s life would be so much easier.

**Reply by fdncred** [reply]
I wouldn't be opposed to respecting `XDG_CONFIG_HOME` env var if it were set. I'm not sure how the other maintainers feel.

**Reply by flying-sheep** [reply]
I’m happy as long as the standard dirs for each platform are used unless explicitly overridden. (So I’m happy now. I’d be unhappy if we changed this to e.g. use XDG default dirs like `~/.config` on macOS or Windows)

**Reply by fdncred** [reply]
> I’m happy as long as the standard dirs for each platform are used unless explicitly overridden.

Yes, that's what I'm saying I'm not opposed to.

**Comment by jeffreyguenther** [comment]
As a macOS user using command-line tools, I was completely surprised to discover that the config was stored in `Application Support`. My expectation was that I would have a `~/.config.nu` or a `~/.config/nushell/*` with all the programs data, or even a `~/.nushell` directory in the extreme.

I would make the case that:
* config.nu
* env.nu
* history.txt
* login.nu
* plugin.nu

And the whole plugin architecture should look in this `.config/nushell` directory.

I don't know this for certain, but I would guess that most macOS users expect a more Linux style directory structure with command-line programs. I have _never_ looked to Application Support for one. Always something in my immediate home directory.

**Reply by flying-sheep** [reply]
> I don't know this for certain, but I would guess that most macOS users expect a more Linux style directory structure with command-line programs.

If that’s true, it’s a sad state of affairs. There’s perfectly standard places for these things, and apparently enough devs were too lazy to use them to the point that people expect things to be in the standard place for *another platform*.

I wish things like [Python’s `platformdirs`](https://pypi.org/project/platformdirs/) and [Rust’s `directories`](https://crates.io/crates/directories) had existed and been used from the first day, then expectations wouldn’t be all over the place.

**Reply by eugenesvk** [reply]
You wish wouldn't help as your very first link is bugged on macOS
https://github.com/platformdirs/platformdirs/issues/98
The OS defaults are not great (read the guidelines, there aren't really "perfectly standard places for these things"), and neither are the cross-platform solutions, so there is not magic wand to get us out of this mess

**Reply by eugenesvk** [reply]
and the second linked library had exactly the same issue!!! https://github.com/dirs-dev/directories-rs/issues/62 that was fixed with an extra confusing mess of `preference_dir` vs `config_dir`!

**Reply by flying-sheep** [reply]
On a similar but separate note: `history.txt` doesn’t contain configuration and therefore doesn’t belong into the config directory. There’s a standard data directory on all platforms that can be used for that.

And since `nushell` is a good tool, it has all the paths readily available so there’s no confusion:

```console
$ $nu | transpose name | where name =~ '.*-path' | transpose -r | into record
╭─────────────────┬────────────────────────────────────────╮
│ config-path     │ /home/phil/.config/nushell/config.nu   │
│ env-path        │ /home/phil/.config/nushell/env.nu      │
│ history-path    │ /home/phil/.config/nushell/history.txt │
│ loginshell-path │ /home/phil/.config/nushell/login.nu    │
│ plugin-path     │ /home/phil/.config/nushell/plugin.nu   │
│ home-path       │ /home/phil                             │
│ temp-path       │ /tmp                                   │
╰─────────────────┴────────────────────────────────────────╯
```

**Reply by flying-sheep** [reply]
@eugenesvk `directories` has 4 major versions, that means that it needed 3 revisions to settle on a design that fully complies to all standards, because things are hard. There seems to be no problem identified with v4, so it seems to very much be the library I would wish for.

The OS defaults aren’t great, but they’re far superior to every project doing its own things.

If everyone could just use the standards places, I could do e.g. this:

- Delete the standard cache dirs to clear all ephemeral disk space
- Use version control on the standard config directory to have configuration synced for all applications cross platform

I’m happy to use Linux because here at least it’s clear that everyone is supposed to follow the XDG spec, as opposed to macOS where there’s apparently no clear guidelines to where stuff goes.

**Reply by eugenesvk** [reply]
> There seems to be no problem identified with v4, so it seems to very much be the library I would wish for.

One problem is the one I mentioned above (confusing duplication of setting variable names), the wrong solution should be phased out, not being kept forever
Another problem is the lack of support of `$XDG_CONFIG_HOME` mentioned here https://github.com/dirs-dev/directories-rs/issues/47
Then there are problems that I'm just not aware of :)

> Use version control on the standard config directory to have configuration synced for all applications cross platform

You can't do that on `~Library/Application Support/`, it's has ~1G folders with a gazillion of files for some apps (so it'd be a chore to exclude) and has all the junk that you'd never need to backup for most of the apps! And also a lot of config files have some constantly changing values that you might not care about. Or history, which you might want to backup, but not sync with other platforms (though these issues aren't fixed automatically with a different folder)
The OS defaults are just bad, so they are not superior to some projects doing the better thing, and definitely not superior to some projects doing the best thing :)

**Reply by norman784** [reply]
I just stop using nushell, because it does not fit my needs, it works pretty good, but for me was a bummer to link files from `~/~Library/Application Support/` to `./config/` with all the breaking changes in each version, also it seems that the maintainers just want to be "correct" in each platform even if every other cli tool in macOS uses XGD paths.

But it is what it is, what I can see is that the maintainers are not using macOS, so for them as users is not an issue.

**Reply by fdncred** [reply]
@norman784 at least 3 of the maintainers use macOS. I use it every day and it's not an issue. 

All it takes is someone to submit a working PR that respects XDG_CONFIG_HOME. I'm pretty sure we'd accept it. I started writing one but didn't have time to figure out how to consolidate the changes properly. I don't fully understand what to do when there is no root shell and nushell is the login shell, where do the env vars come from then?

**Reply by norman784** [reply]
> All it takes is someone to submit a working PR that respects XDG_CONFIG_HOME. I'm pretty sure we'd accept it. 

@fdncred good to know that the project is willing to support XDG.

>  I don't fully understand what to do when there is no root shell and nushell is the login shell, where do the env vars come from then?

Do you know how it works on linux or other *nix OSes (besides macOS)?

**Reply by flying-sheep** [reply]
XDG is a standard for freedesktop platforms, of which macOS isn’t one. I think mixing standards makes things more confusing and less predictable. However, I guess if people define `$env.XDG_CONFIG_HOME` on macOS, they already know that they’re using tools that don’t follow platforms standards, so idk.

I’d personally define tool specific env variables (that of course work on any platform) and rely on `platformdirs` handling things like the XDG standard. Which it perfectly does. On XDG platforms.

**Reply by fdncred** [reply]
> Do you know how it works on linux or other *nix OSes (besides macOS)?

Nope

**Reply by eugenesvk** [reply]
> I don't fully understand what to do when there is no root shell and nushell is the login shell, where do the env vars come from then?

they can come from `~/Library/LaunchAgents/environment.plist` as described in this [SO answer](https://stackoverflow.com/a/36161817), so you could add this advice to the nushell guide, telling users how they can set their XDG env vars for a login shell
Also, given that `~/.config` is the most common location, even without an explicit `XDG_CONFIG_HOME` variable you could check if that folder has a folder with nushell configs, and if it does, use them.


↓ sets `XDG_CONFIG_HOME` to `~/.config` and should be available to a login nushell 
```xml
<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd"><plist version="1.0">
<dict>
<key>Label</key><string>my.startup.shell_agnostic.environment</string>
<key>ProgramArguments</key><array><string>sh</string><string>-c</string><string>
  launchctl setenv XDG_CONFIG_HOME ~/.config
</string></array>
<key>RunAtLoad</key><true/>
</dict>
</plist>
```

**Reply by jeffreyguenther** [reply]
I think this is a great option. 

Providing some documentation why this change is necessary and when in the installation process to do it would be helpful. I've never edited a plist in my decade plus of using macOS, so extra context would be helpful.
